### DISCUSSION TITLE
Add xml body nodes to each request

### DISCUSSION THREAD

**Discussion by rico** [phpro/soap-client#465] [Q&A] [answered]
This library was very easy to get started, but started to get overwhelming pretty quick :-)

Can somebody put me in the right direction on how I would add two nodes to the body for all the requests? A custom middleware? How do I get started with these?

Thank you all.

**Comment by veewee** [comment, marked as answer]
Hello @rico ,

I can imagine it gets overwhelming... That's the downside of having it ultra flexible. :)
We do our best to document everything as good as possible, but feel free to improve the documentation.

There are a few ways to do this depending on your exact case:

1. If you have to pass e.g. a token to every request, and this token is part of the generated soap types  - then you can use an event listener to fill in this data at object level.
2. If it's more like a soap-header that needs to be added, then it can be done trough a HTTP middleware.


**If you think an event subscriber is the way to go:**

The basics are covered here: https://github.com/phpro/soap-client/blob/master/docs/events.md

We use this approach for adding e.g. a session identifier to every soap request.
The way we use it:

* Create a custom rule that determines that the type has the property to link the token to.
* Add an interface to this model - this interface can be used by the listener to determine if the token needs to be added.
* Create a listener that registers the token on the model

Example model implementation:

```php
final class RegisterSessionListener
{
    public function __construct(private SessionProviderInterface $sessionProvider)
    {
    }

    public function onClientRequest(RequestEvent $event): void
    {
        $request = $event->getRequest();
        if (!$request instanceof RequiresShessionInterface) {
            return;
        }

        $event->registerRequest(
            $request->withSession(
                $this->sessionProvider->provide()
            )
        );
    }
}
```




**If you think an HTTP middleware is the way to go:**

* https://github.com/php-soap/xml : These are some helpers wo make it possible to manipulate the xml of an HTTP request.
* https://github.com/php-soap/psr18-transport#middleware: you can take a look at the existing middleware, most likely you're interested in this one: https://github.com/php-soap/psr18-transport/blob/main/src/Middleware/SoapHeaderMiddleware.php of which you can start to manipulate the soap body. It uses the xml helpers linked above.

Example middleware for manipulating soap body

```php
    public function handleRequest(RequestInterface $request, callable $next, callable $first): Promise
    {
        return $next((new XmlMessageManipulator)(
            $request,
            function (Document $document) {
                $body = $document->locate(new SoapBodyLocator());
                append($YOURELEMENTS)($body);
                
                return $document;
            }
        ));
    }
```

You can read more about manipulating the document here: https://github.com/veewee/xml/blob/main/docs/dom.md


Hope this gets you in the right direction :)

**Comment by rico** [comment]
Hi @veewee, a huge thank you for the detailed and helpful answer. I was able to solve it using a simple event subscriber.

I guess this project is almost too simple to get started, so that people like me, without much of knowledge in the domain, get stuck easily. But your answer is great to find the right direction :-)
